castoredc_api/importer/async_helpers.py

```diff
@@ -0,0 +1,235 @@
+    for idx, chunk in enumerate(chunks):
+        async with httpx.AsyncClient(
+            headers=study.client.headers,
+            timeout=study.client.timeout,
+            limits=study.client.limits,
+        ) as client:
+            tasks = [async_upload_report_data(item, client, study) for item in chunk]
+
+            # Show progress bar while handling responses
+            temp_responses = [
+                await response
+                for response in tqdm(
+                    asyncio.as_completed(tasks),
+                    total=len(tasks),
+                    desc=f"Async Uploading {idx + 1}/{len(chunks)}",
+                    file=sys.stdout,
+                )
+            ]
+            responses = responses + temp_responses
+    return responses
+
+
+async def async_upload_report_data(item, client, study):
+    """Coroutine to upload a single row of report data to a new report and handle the response."""
+    # Copy because we don't want to overwrite dict
+    feedback = copy.deepcopy(item["row"])
+    try:
+        instance = await async_create_report_instance(
+            report_id=item["report_id"],
+            record_id=item["row"]["record_id"],
+            report_name_custom=item["report_name"],
+            study=study,
+            client=client,
+        )
+        body = create_report_body(instance, item["row"], study, item["upload_datetime"])
+        url = (
+            study.client.study_url
+            + f"/record/{item['row']['record_id']}/data-point-collection/"
+            f"report-instance/{instance['id']}"
+        )
+        json = {"data": body, "common": item["common"]}
+        feedback = await async_upload_data(client, feedback, json, study, url)
+    except httpx.HTTPStatusError as error:
+        feedback["error"] = error.response.json()
+    return feedback
+
+
+async def async_create_report_instance(
+    report_id: str,
+    record_id: str,
+    report_name_custom: str,
+    study: "CastorStudy",
+    client: httpx.AsyncClient,
+):
+    """Creates a survey package instance asynchronously."""
+    url = study.client.study_url + f"/record/{record_id}/report-instance"
+    body = {
+        "report_id": report_id,
+        "report_name_custom": report_name_custom,
+        "parent_id": None,
+    }
+    response = await client.post(url=url, json=body)
+    response.raise_for_status()
+    return response.json()
+
+
+async def async_upload_data(client, feedback, json, study, url):
+    """Sends body to url and awaits response"""
+    response = await client.post(url=url, json=json)
+    try:
+        response.raise_for_status()
+        formatted_response = format_feedback(response.json(), study)
+        feedback["error"] = None
+        feedback["success"] = formatted_response["success"]
+        feedback["failed"] = formatted_response["failed"]
+    except HTTPStatusError as error:
+        feedback["error"] = error.response.json()
```

Comment:
I've made this comment elsewhere as well, but the response may not be json-decodable. That would blow up here.

Reply:
Resolving this conversation as it is a duplicate of other conversations.